src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,205 @@
+
+        try {
+            tasks = new TaskList(storage.loadFromDisk());
+            repeatUntilQuit();
+            exit();
+        } catch (DukeException ex) {
+            System.out.println(ex.getMessage());
+        }
+    }
+
+    private void exit() {
+        ui.printGoodbyeMessage();
+    }
+
+    private void repeatUntilQuit() {
+
+        while (true) {
+            String input = ui.getUserInput();
+            String[] words = parser.parseInput(input);
+            String keyword = parser.getKeyword(input);
+
+            try {
+                switch (keyword) {
+                    case "bye":
+                        return;
+                    case "list":
+                        ui.printAllTasks(tasks.getTaskList());
+                        break;
+                    case "todo":
+                        createToDos(words);
+                        break;
+                    case "deadline":
+                        createDeadlines(words);
+                        break;
+                    case "event":
+                        createEvents(words);
+                        break;
+                    case "mark":
+                        mark(words);
+                        break;
+                    case "unmark":
+                        unmark(words);
+                        break;
+                    case "delete":
+                        deleteTask(words);
+                        break;
+                    case "find":
+                        findTask(words);
+                        break;
+                    default:
+                        throw new DukeInvalidException();
+                }
+            } catch (DukeException ex) {
+                System.out.println(ex.getMessage());
+            }
+        }
+    }
+
+    private void mark(String[] input) throws DukeException {
+        if (input.length < 2) {
+            throw new DukeEmptyException(input[0]);
+        }
+        try {
+            int taskIndex = parser.getTaskIndex(input);
+            Task markedTask = tasks.getTask(taskIndex);
+            markedTask.markAsDone();
+            ui.printTaskMarked(markedTask);
+            storage.saveToDisk(tasks.getTaskList());
+        } catch (NumberFormatException | IndexOutOfBoundsException ex) {
+            throw new DukeIndexOutOfBoundsException(tasks.getSize());
+        }
+    }
+
+    private void unmark(String[] input) throws DukeException {
+        if (input.length < 2) {
+            throw new DukeEmptyException(input[0]);
+        }
+        try {
+            int taskIndex = parser.getTaskIndex(input);
+            Task unmarkedTask = tasks.getTask(taskIndex);
+            unmarkedTask.unmarkAsNotDone();
+            ui.printTaskUnmarked(unmarkedTask);
+            storage.saveToDisk(tasks.getTaskList());
+        } catch (NumberFormatException | IndexOutOfBoundsException ex) {
+            throw new DukeIndexOutOfBoundsException(tasks.getSize());
+        }
+    }
```

Comment:
It seems like `Duke::mark` and `Duke::unmark` perform roughly the same procedure with only a differing call to the Ui object. Perhaps we may have a more general function that abstracts out the identical lines, taking in a boolean that decide whether to mark or unmark, such that the code can be made more compact?
```suggestion
    ...
    case "mark":
        annotate(words, true);
        break;
     case "unmark":
        annotate(words, false);
        break;
    ....
    
    private void annotate(String[] input, boolean isMarked) throws DukeException {
        if (input.length < 2) {
            throw new DukeEmptyException(input[0]);
        }
        try {
            int taskIndex = parser.getTaskIndex(input);
            Task annotatedTask= tasks.getTask(taskIndex);
            if (isMarked) {
                annotatedTask.markAsDone();
                ui.printTaskMarked(annotatedTask);
            } else {
                annotatedTask.unmarkAsNotDone();
                ui.printTaskUnmarked(annotatedTask);
            }
            storage.saveToDisk(tasks.getTaskList());
        } catch (NumberFormatException | IndexOutOfBoundsException ex) {
            throw new DukeIndexOutOfBoundsException(tasks.getSize());
        }
    }
```